### PR TITLE
fix: grpc reconnection 

### PR DESCRIPTION
**Pull request by kstasik**
I have successfully reproduced this issue on 0.15.0 but i think it also applies to recent version. After disconnection it tries to reconnect only once. With this fix it should try reconnect periodically until successful connection.

Please let me know what do you think. Maybe there is some other better way to fix it.

### CODE REVIEW

**Comment by MrAlias** [comment]
Is there an issue tracking this? If not, can we start there? It would be nice to better understand the problem this is attempting to solve.

**Comment by kstasik** [comment]
I hope it helps a little bit:

https://github.com/open-telemetry/opentelemetry-go/issues/1524

**Comment by jmacd** [comment]
It would be great if we had a test that would prevent regressions, since this bug has been a major issue of this nature. I am not very familiar with the code, but this certainly looks correct.

**Review by jmacd — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,10 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Rename project default branch from `master` to `main`.
 - Reverse order in which `Resource` attributes are merged, per change in spec. (#1501)
 
+## Fixed
+
+- Fixed otlpgrpc reconnection issue.
```
Comment (XSAM) [inline CHANGELOG.md:18]:
```suggestion
- Fixed otlpgrpc reconnection issue. (#1521)
```

**Review by XSAM — COMMENTED** [review]
I can't reproduce this bug, and the test `TestNewExporter_collectorConnectionDiesThenReconnects` works fine.

File: exporters/otlp/otlpgrpc/connection.go [hunk 2]
```diff
@@ -135,21 +135,23 @@ func (c *connection) indefiniteBackgroundConnection() {
...
 		// 2. Otherwise block until we are disconnected, and
 		//    then retry connecting
-		select {
-		case <-c.stopCh:
-			return
-
-		case <-c.disconnectedCh:
-			// Quickly check if we haven't stopped at the
-			// same time.
+		if c.connected() {
```
Comment (XSAM) [inline exporters/otlp/otlpgrpc/connection.go:138]: This might cause infinitely retry even the connection is connected.

**Comment by kstasik** [comment]
thank you - i didn't notice integration tests. I'm not sure about my solution but I think this test doesn't cover this line:

https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/otlpgrpc/connection.go#L158

Which i think causes goroutine to hang. I will try to write new test or extend this.

**Comment by kstasik** [comment]
I think changing disconnectedCh to buffered may better option but I will change this PR once i have time to with integration tests.

**Review by punya — APPROVED** [review]
This may also fix the test flake I ran into elsewhere (TestNewExporter_collectorConnectionDiesThenReconnects).

**Comment by punya** [comment]
> I can't reproduce this bug, and the test `TestNewExporter_collectorConnectionDiesThenReconnects` works fine.

Actually, I saw it flake several times in CI yesterday, and locally I see it fail about 5% of the time (https://gist.githubusercontent.com/punya/1db2a5fc7aede3baa06685997e145f3b/raw/bc52d8e2ba68772539a36c87f8ff9d1492e2ecf6/runs.txt).

**Comment by kstasik** [comment]
btw. in this case connected() function was very misleading for me (grpc exporter). It doesn't mean exporter is connected and thanks to this reconnection goroutine it's almost always returns true.

First reconnection always successfully creates new "connection" which is really not established.
https://pkg.go.dev/google.golang.org/grpc#DialContext

I changed fix to buffered channel, please review and let me know what do you think. Does it may cause any other issues?

**Review by XSAM — APPROVED** [review]
Looks good 👍

**Review by Aneurysm9 — APPROVED** [review]

**Review by codeboten — APPROVED** [review]

**Pull request merged**